fix(datasets): Various fixes

astrojuanlu opened the pull request:
## Description
<!-- Why was this PR created? -->

## Development notes
<!-- What have you changed, and how has this been tested? -->

## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

astrojuanlu commented:
Went from 362 seconds to 193 seconds @lrcouto. I cherry-picked @merelcht 's commit to pin Dask too. And the docs built successfully:

https://beta.readthedocs.org/projects/kedro-datasets/builds/23742913/

Feel free to take it over from here.

lrcouto commented:
Are these changes related to the unit-tests failing? I'm trying to reproduce the error locally here but my local environment is not too trustworthy.

astrojuanlu commented:
Unsure what's the hell is going on, but I attempted several fixes. The one that needs careful review 🔍⚠️  is the Tensorflow dataset changes I made.

- Use `uv` on RTD
- Pin Dask, both in test and in the requirements proper, because of #609 (proper term fix is allow pandas 2 for testing)
- Use smaller Hugging Face model to help the test pass + Use PyTorch-based model instead of TensorFlow-based one
- ⚠️ Rework TensorFlow dataset to work with Keras 3 (which deprecated `save_format` and relies on file extensions only) I didn't attempt to fix the `.h5` format so I'm expecting some tests to properly fail but at least the doctests should pass

I'm leaving this here

See #600

astrojuanlu commented:
Managed to unblock RTD and the doctests

merelcht commented on kedro-datasets/kedro_datasets/tensorflow/tensorflow_model_dataset.py:
```diff
@@ -134,12 +134,11 @@ def __init__(  # noqa: PLR0913
-                self._fs.copy(load_path, path)
-            else:
-                self._fs.get(load_path + "/", path, recursive=True)
+                raise NotImplementedError("Juanlu is on a rush")
```
😂

merelcht reviewed (commented):
I'm happy with most of these changes, but the Tensorflow ones seem a bit more involved. Changing the format from `h5` to `keras` seems breaking, so if this goes in we should immediately follow up on getting a proper fix before any release.

astrojuanlu commented:
Indeed 👍🏽 I think to be completely honest the change to `.keras` extension is unwarranted. The culprit seems to be that we assume `.h5` are files and otherwise they're directories, and that trips tensorflow up, don't know why. The removal of the default save arguments is required though.

astrojuanlu commented on kedro-datasets/tests/tensorflow/test_tensorflow_model_dataset.py:
```diff
@@ -193,6 +193,8 @@ def test_unused_subclass_model_hdf5_save_format(
     ):
         """Test TensorFlowModelDataset cannot save subclassed user models in HDF5 format
 
+        FIXME: What is this test doing??
```
FYI

astrojuanlu commented:
Looks like I fixed all the tests, at least on Ubuntu.

@DimedS @noklam @merelcht I removed a test that was expecting an error that is not erroring anymore, and removed an expected exception from another test that is not happening. That in principle looks like good news to me but it requires another pair of eyes.

I leave this to you now.

astrojuanlu commented on kedro-datasets/.readthedocs.yaml:
```diff
@@ -11,14 +11,10 @@ build:
-    # These are technically installation steps, due to RTD's limit we need to inject the installation earlier.
-      - python -m pip install --upgrade --no-cache-dir sphinx readthedocs-sphinx-ext
-      - python -m pip install --upgrade --upgrade-strategy only-if-needed --no-cache-dir ./kedro-datasets[docs,test]
+      - python -m pip install "uv==0.1.13"
```
`uv` is at a newer version now, but "if it works, don't touch it"...

astrojuanlu commented on kedro-datasets/kedro_datasets/tensorflow/tensorflow_model_dataset.py:
```diff
@@ -62,7 +63,7 @@ class TensorFlowModelDataset(AbstractVersionedDataset[tf.keras.Model, tf.keras.M
     """
 
     DEFAULT_LOAD_ARGS: dict[str, Any] = {}
-    DEFAULT_SAVE_ARGS: dict[str, Any] = {"save_format": "tf"}
```
This is required for Keras 3, it's also removed from the tests

astrojuanlu commented on kedro-datasets/pyproject.toml:
```diff
@@ -33,7 +33,7 @@ api = ["kedro-datasets[api-apidataset]"]
 biosequence = ["kedro-datasets[biosequence-biosequencedataset]"]
 
-dask-parquetdataset = ["dask[complete]>=2021.10", "triad>=0.6.7, <1.0"]
+dask-parquetdataset = ["dask[complete]>=2021.10, <2024.3", "triad>=0.6.7, <1.0"]
```
Probably we should add a comment here referencing #609

astrojuanlu commented on kedro-datasets/pyproject.toml:
```diff
@@ -232,7 +232,7 @@ test = [
     "datasets",
     "huggingface_hub",
-    "transformers",
+    "transformers[torch]",
```
Previously `transformers` (Hugging Face) was using TensorFlow (since it's installed as part of the tests), but PyTorch support is more robust

astrojuanlu commented on kedro-datasets/tests/tensorflow/test_tensorflow_model_dataset.py:
```diff
@@ -408,18 +402,6 @@ def test_version_str_repr(self, tf_model_dataset, versioned_tf_model_dataset):
-        self, tf_model_dataset, versioned_tf_model_dataset, dummy_tf_base_model
-    ):
-        """Check behavior when attempting to save a versioned dataset on top of an
-        already existing (non-versioned) dataset. Note: because TensorFlowModelDataset
-        saves to a directory even if non-versioned, an error is not expected."""
```
Now we're not saving to a directory anymore

DimedS reviewed (approved):
Many thanks, @astrojuanlu . It looks good to me at first glance. I'll take a closer look at it today. For now, I'm approving it to unblock the PRs.

SajidAlamQB reviewed (approved):
This looks great, thank you @astrojuanlu!

astrojuanlu commented:
This is going in
